File: src/main/java/io/github/takenoha/towerdefense/tactical/TacticalBuildLifecycle.java
package io.github.takenoha.towerdefense.tactical;

import java.util.UUID;

/** Idempotent lifecycle boundary called by the defense runtime. */
public interface TacticalBuildLifecycle {
    TacticalUnlockResult activateAtPreparation(UUID defenseId, UUID operationId);

    TacticalUnlockResult advanceAfterWave(
            UUID defenseId,
            int completedWaveCount,
            int totalWaveCount,
            UUID operationId);

    TacticalUnlockResult activateFinalTier(UUID defenseId, UUID operationId);

    void markTerminal(UUID defenseId, TacticalTerminalResult result, UUID operationId);
}

File: src/main/kotlin/io/github/takenoha/towerdefense/tactical/TacticalBuildLifecycle.kt
package io.github.takenoha.towerdefense.tactical

import java.util.UUID

/** Idempotent lifecycle boundary called by the defense runtime. */
interface TacticalBuildLifecycle {
    fun activateAtPreparation(defenseId: UUID, operationId: UUID): TacticalUnlockResult

    fun advanceAfterWave(
        defenseId: UUID,
        completedWaveCount: Int,
        totalWaveCount: Int,
        operationId: UUID,
    ): TacticalUnlockResult

    fun activateFinalTier(defenseId: UUID, operationId: UUID): TacticalUnlockResult

    fun markTerminal(
        defenseId: UUID,
        result: TacticalTerminalResult,
        operationId: UUID,
    )
}
